content/en/blog/_posts/2021-08-11-memory-qos-cgroups-v2.md

```diff
@@ -0,0 +1,73 @@
+
+It may cause:
+- Pod / Container memory requests can't be fully reserved, page cache is at risk of being recycled
+- Pod / Container memory allocation is not well protected, there may occur allocation latency frequently when node memory nearly runs out
+- Memory overcommit of container is not throttled, there may increase risk of node memory pressure
+
+Kubernetes 1.22 brings Memory QoS to alpha to pay attention on the quality of service of memory resources. It uses cgroup v2 new capabilities `memory.min / memory.high` in memory controller to retain memory requests and throttle overcommit limits. This feature can improve memory availability and reduce memory requests reclamation, which is of great significance to ensure the stability of both workloads and nodes.
+
+## How does it work?
+It uses two interfaces `memory.min / memory.high` of cgroup v2 memory controller to work.
+| File | Description |
+| -------- | -------- |
+| memory.min | memory.min specifies a minimum amount of memory the cgroup must always retain, i.e., memory that can never be reclaimed by the system. If the cgroup's memory usage reaches this low limit and can’t be increased, the system OOM killer will be invoked. **We map it to `requests.memory`.** |
+| memory.high | memory.high is the memory usage throttle limit. This is the main mechanism to control a cgroup's memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is max, meaning there is no limit. **We use a formula to calculate `memory.high` depending on `limits.memory/node allocatable memory` and a memory throttling factor.** |
+
+Both container and pod are considered. When cgroup v2 unified and MemoryQoS feature gate are both enabled, kubelet sets `requests.memory` to `memory.min` at container / pod level cgroup. And it also sets `memory.high` to throttle container memory overcommit allocation. 
+
+![](https://i.imgur.com/FpR0cUK.png)
+
+The formula is as follows:
+```
+// Container
+/cgroup2/kubepods/pod<UID>/<container-id>/memory.min=pod.spec.containers[i].resources.requests[memory]
+/cgroup2/kubepods/pod<UID>/<container-id>/memory.high=(pod.spec.containers[i].resources.limits[memory]/node allocatable memory)*memory throttling factor // Burstable
+// Pod
+/cgroup2/kubepods/pod<UID>/memory.min=sum(pod.spec.containers[i].resources.requests[memory])
+// QoS ancestor cgroup
+/cgroup2/kubepods/burstable/memory.min=sum(pod[i].spec.containers[j].resources.requests[memory])
+
+```
+
+Cgroup v2 memory controller would retain memory requests or throttle memory usages in the pod or container cgroup based on the set value. Because effective min boundary is limited by `memory.min` values of all ancestor cgroups, kubelet also sets qos or root cgroup if necessary.
```

Comment:
I'm afraid I'm not sure what point is being made here.
Is one of these true:
```suggestion
The cgroup v2 memory controller enforces memory requests and ceilings
per-pod or per-container. If you set. The design of cgroup v2 also implies
for the `memory.min` behavior to be as described, all ancestor cgroups
must have appropriate `memory.min` values. To ensure this, the kubelet sets
`memory.min` on the parent cgroup (representing the Pod).
```
```suggestion
The cgroup v2 memory controller enforces memory requests and ceilings
per-pod or per-container. If you set. The design of cgroup v2 also implies
for the `memory.min` behavior to be as described, all ancestor cgroups
must have appropriate `memory.min` values. To ensure this, the kubelet sets
`memory.min` on all ancestor cgroups up to the root.
You are responsible for ensuring that your init system (or other operating system
level tooling) does not also try to manage `memory.min` on those ancestor cgroups.
```
?

Also, should we mention `memory_recursiveprot` as a mount option?

Reply:
I think the second is betther.